### ISSUE TITLE
import ctf issue

### ISSUE THREAD

**Issue by azazellochg** [I2PC/scipion#588]
1. import micrographs with names img1.mrc, img2.mrc ... img3050.mrc
2. put in one folder results from gctf or ctffind3/4 i.e. for a single mic: img21.ctf, img21_EPA.log, img21_ctffind3.log
3. run import ctf protocol with pattern "*ctffind3.log" - 1 matching file

I get ctfs and mics with 2 (!) items each, for img2 and img21. CTF values are the same, but micrographs are correct. I have added few prints into ctf import code and the output gives:

```
00012:   Matching files:  1
00013:   CTF for micrograph id 1 was not found. Removed from set of micrographs.
...
01124:   fileId= None mic.getObjId()= 1112 fileName= /home/igbmc/gsharov/scratch/test/img21_ctffind3.log micName= img2.mrc micBase= img2
...
01236:   fileId= None mic.getObjId()= 1224 fileName= /home/igbmc/gsharov/scratch/test/img21_ctffind3.log micName= img21.mrc micBase= img21
...
```

The problem is that [here](https://github.com/azazellochg/scipion/blob/devel/pyworkflow/em/protocol/protocol_import/ctfs.py#L126) in `micBase in fileName or micName in fileName` the first condition returns true, when we compare **img2** with **/home/igbmc/gsharov/scratch/test/img21_ctffind3.log**

**Comment by delarosatrevin** [comment]
What is the state of this issue?

**Comment by yaizar** [comment]
It hasn't been worked on as far as I know, I'll start taking a look probably next week

**Comment by yaizar** [comment]
Been looking into this, unfortunately I don't think this is an issue we can solve without compromising the protocol _genericness_ and simplicity... Either we need the user to tell us **exactly** where in the pattern can we find the mic name we need to match (so we keep the input filename generic, but we make the protocol input more difficult). Or we'd need to match the mic name with the folder name of the ctf we are importing (which is consistent with Scipion's way of saving ctfs, but we'd lose genericness and users who are importing their ctfs from somewhere else should put them in folders too. not ideal). 
So as for a good solution to make this work seamlessly I personally think we're out of luck (suggestions welcome!).. but one thing we can do is validate the output (i.e. we need to have the same number of ctf files matched and micrographs) and if the number is not the same (like in this case) we make the protocol fail or show a warning.

**Comment by delarosatrevin** [comment]
I think the work around in this PR still does not fix the import problem. I have found a couple of problems:
1) If no matching, empty sets are created (probably this happened before the changes of this PR)
2) The is not matching of all items for names like: mic1.mrc, mic2.mrc and mic10.mrc

I have added a failing test case for it.
Still, I think there is no easy solution for this problem without requiring from the user a more string naming matching (now is more flexible and allow partially names matching). One possible solution could be to consider longer names first, but still some wrong match could happens if not all CTF files are present for import.

**Comment by yaizar** [comment]
I think the work around in this PR fixes the import problem specified in this issue, but not all import problems:
1. This is accurate, creates empty sets without this PR's code. This is unrelated to having files with conflicting names. Should we do something about it? Fail instead of creating empty set? 
2.  I don't understand this :/ can you rephrase?

I have checked your test, and the used pattern (mic*/*txt) returns 6 matched files for 3 micrographs, eg: mic1 would match  mic1/ctfEstimation.txt and mic1/ctfEstimation_avrot.txt, so I think this problem is more of a bad pattern choice than a problem with this pr's code.  I think the normal pattern for this case would be mic*/ctfEstimation.txt - and with that it works good. Now it is true that we can change the logic about what to do when the user enters a bad pattern and there are more files matched than micrographs, I suggest adding a warning and trying them all until there is a successful import. Other ideas?

**Comment by azazellochg** [comment]
could we just ignore *_avrot.txt files? "_avrot" is a constant suffix added by ctffind4 and cannot be changed by user.

**Comment by yaizar** [comment]
I personally think it is best to deal with that by choosing the right pattern rather than coding it inside the protocol, but if you guys agree on that then sure, we can make it happen.

**Comment by delarosatrevin** [comment]
Hi @yaizar, you were right, the fail of the test that I added was somehow related to the pattern used. Nonetheless, we should try make the software as robust as possible (until some extent, of course). Anyway, it seems that with the last commit is also work properly even with the wrong pattern, thanks!

@azazellochg I agree with Yaizar that I prefer to document and encourage to use the correct pattern for CTF files. Can you test if it is working now for the case where you detected the problem?

After that I think the PR is ready to be merged.

**Comment by azazellochg** [comment]
Hi @delarosatrevin @yaizar 
 I ran a test with both gctf/ctffind4, the matching works and nice warning are displayed (using default pattern img*.txt or *gctf.log). 

About the pattern use: I agree that user should provide a more specific pattern for ctf files. In this particular case (ctffind4), we want to avoid *avrot files. I tried to provide "img*[^avrot].txt" pattern but it does not work.
